### PR TITLE
Plugin for cookie manager

### PR DESCRIPTION
**Pull request by rjharishabh**
Cookie Manager Plugin

### CODE REVIEW

File: administrator/components/com_admin/sql/updates/mysql/4.1.0-2021-06-08.sql [hunk 1]
```diff
@@ -24,3 +24,6 @@ CREATE TABLE IF NOT EXISTS `#__cookiemanager_cookies` (
 
 INSERT INTO `#__extensions` (`package_id`, `name`, `type`, `element`, `folder`, `client_id`, `enabled`, `access`, `protected`, `locked`, `manifest_cache`, `params`, `custom_data`) VALUES
 (0, 'com_cookiemanager', 'component', 'com_cookiemanager', '', 1, 1, 1, 0, 1, '', '{"policylink":"","position":"bottom"}', '');
+
+INSERT INTO `#__extensions` (`package_id`, `name`, `type`, `element`, `folder`, `client_id`, `enabled`, `access`, `protected`, `manifest_cache`, `params`, `custom_data`, `checked_out`, `checked_out_time`, `ordering`, `state`) VALUES
+(0, 'plg_system_cookiemanager', 'plugin', 'cookiemanager', 'system', 0, 1, 1, 0, '', '', '', 0, '0000-00-00 00:00:00', 0, 0);
```
Comment (richard67) [inline administrator/components/com_admin/sql/updates/mysql/4.1.0-2021-06-08.sql]: As far as I remember I had explained you once that when looking for examples to insert new extensions, you should not take too old scripts as example because they run before the null dates changes.

But you caught such an old example where 0 instead of NULL is used for `checked_out`, and '0000-00-00 00:00:00' instead of NULL is used for `checked_out_time`.

In addition it is just silly to have these 2 insert statements into the same table below each other using different columns in the statement, depending where they have been copied from.

You should instead use the first insert statement to insert both records at once, the one for the component and the other one for the plugin.

So it would be:

```
INSERT INTO `#__extensions` (`package_id`, `name`, `type`, `element`, `folder`, `client_id`, `enabled`, `access`, `protected`, `locked`, `manifest_cache`, `params`, `custom_data`) VALUES
(0, 'com_cookiemanager', 'component', 'com_cookiemanager', '', 1, 1, 1, 0, 1, '', '{"policylink":"","position":"bottom"}', ''),
(0, 'plg_system_cookiemanager', 'plugin', 'cookiemanager', 'system', 0, 1, 1, 0, 1, '', '', '');
```

File: administrator/components/com_admin/sql/updates/postgresql/4.1.0-2021-06-08.sql [hunk 2]
```diff
@@ -25,3 +25,6 @@ CREATE INDEX "#__cookiemanager_cookies_idx_createdby" on "#__cookiemanager_cooki
 
 INSERT INTO "#__extensions" ("package_id", "name", "type", "element", "folder", "client_id", "enabled", "access", "protected", "locked", "manifest_cache", "params", "custom_data", "ordering", "state") VALUES
 (0, 'com_cookiemanager', 'component', 'com_cookiemanager', '', 1, 1, 1, 0, 1, '', '{"policylink":"","position":"bottom"}', '', 0, 0);
+
+INSERT INTO "#__extensions" ("package_id", "name", "type", "element", "folder", "client_id", "enabled", "access", "protected", "manifest_cache", "params", "custom_data", "checked_out", "checked_out_time", "ordering", "state") VALUES
+(0, 'plg_system_cookiemanager', 'plugin', 'cookiemanager', 'system', 0, 1, 1, 0, '', '', '', 0, '1970-01-01 00:00:00', 0, 0);
```
Comment (richard67) [inline administrator/components/com_admin/sql/updates/postgresql/4.1.0-2021-06-08.sql]: Same as above for the MySQL update script.

File: administrator/components/com_admin/sql/updates/mysql/4.1.0-2021-06-08.sql (hunk 1 above)
Comment (rjharishabh) [reply administrator/components/com_admin/sql/updates/mysql/4.1.0-2021-06-08.sql]: Thanks and sorry I forgot very quickly

File: installation/sql/mysql/base.sql [hunk 3]
```diff
@@ -323,6 +323,7 @@ INSERT INTO `#__extensions` (`package_id`, `name`, `type`, `element`, `folder`,
 (0, 'plg_system_accessibility', 'plugin', 'accessibility', 'system', 0, 0, 1, 0, 1, '', '{}', '', 1, 0),
 (0, 'plg_system_actionlogs', 'plugin', 'actionlogs', 'system', 0, 1, 1, 0, 1, '', '{}', '', 2, 0),
 (0, 'plg_system_cache', 'plugin', 'cache', 'system', 0, 0, 1, 0, 1, '', '{"browsercache":"0","cachetime":"15"}', '', 3, 0),
+(0, 'plg_system_cookiemanager', 'plugin', 'cookiemanager', 'system', 0, 1, 1, 0, 0, '', '{}', '', 0, 0),
```
Comment (richard67) [inline installation/sql/mysql/base.sql]:
```suggestion
(0, 'plg_system_cookiemanager', 'plugin', 'cookiemanager', 'system', 0, 1, 1, 0, 1, '', '{}', '', 0, 0),
```
You should use the same values here as for the update SQL. The `locked` column should have value 1 because your plugin is a core extension which can be enabled or disabled but it shall not be possible to uninstall it.

File: installation/sql/postgresql/base.sql [hunk 4]
```diff
@@ -329,6 +329,7 @@ INSERT INTO "#__extensions" ("package_id", "name", "type", "element", "folder",
 (0, 'plg_system_accessibility', 'plugin', 'accessibility', 'system', 0, 0, 1, 0, 1, '', '{}', '', 1, 0),
 (0, 'plg_system_actionlogs', 'plugin', 'actionlogs', 'system', 0, 1, 1, 0, 1, '', '{}', '', 2, 0),
 (0, 'plg_system_cache', 'plugin', 'cache', 'system', 0, 0, 1, 0, 1, '', '{"browsercache":"0","cachetime":"15"}', '', 3, 0),
+(0, 'plg_system_cookiemanager', 'plugin', 'cookiemanager', 'system', 0, 1, 1, 0, 0, '', '{}', '', 0, 0),
```
Comment (richard67) [inline installation/sql/postgresql/base.sql]:
```suggestion
(0, 'plg_system_cookiemanager', 'plugin', 'cookiemanager', 'system', 0, 1, 1, 0, 1, '', '{}', '', 0, 0),
```
Same as for the mysql script.

**Comment by richard67** [comment]
@rjharishabh Sorry I have forgotten about it until now, but there is something missing in this PR and the other one. You couldn't know that, and we often forget that too.

There is a list of core extensions here: https://github.com/joomla/joomla-cms/blob/4.0-dev/libraries/src/Extension/ExtensionHelper.php#L49

As you can see it is organized in sections, first components, then files, languages, libraries, modules and plugins.

You need to add your plugin to that (and also the component, so I have commented in the other PR, too).

It is necessary for the backend being able to find it in the extensions manager when filtering for core extensions, and it is necessary when finalizing the installation or update for the installer to update manifest cache in database.

File: installation/sql/mysql/base.sql (hunk 3 above)
Comment (rjharishabh) [reply installation/sql/mysql/base.sql]: Thanks

File: installation/sql/postgresql/base.sql (hunk 4 above)
Comment (rjharishabh) [reply installation/sql/postgresql/base.sql]: thanks

File: plugins/system/cookiemanager/cookiemanager.php [hunk 5]
```diff
@@ -0,0 +1,222 @@
...
+	 * @since  __DEPLOY_VERSION__
+	 */
+	 protected $db;
+
+	/**
+	 * Add assets for the modal.
+	 *
+	 * @return	void
+	 *
+	 * @since	__DEPLOY_VERSION__
```
Comment (richard67) [inline plugins/system/cookiemanager/cookiemanager.php]:
```suggestion
	 * @return  void
	 *
	 * @since   __DEPLOY_VERSION__

```

File: plugins/system/cookiemanager/cookiemanager.php [hunk 6]
```diff
@@ -0,0 +1,222 @@
...
+					]
+				)
+				->order($db->quoteName('lft'));
+
+			$db->setQuery($query);
+			$category = $db->loadObjectList();
+
+			$bannerBody = '<p>' . Text::_('COM_COOKIEMANAGER_COOKIE_BANNER_DESCRIPTION') . '</p><p><a '
+			. ' href="' . $menuitem->link . '">' . Text::_('COM_COOKIEMANAGER_VIEW_COOKIE_POLICY') . '</a></p>'
+			. '<h5>' . Text::_('COM_COOKIEMANAGER_MANAGE_CONSENT_PREFERENCES') . '</h5><ul>';
```
Comment (richard67) [inline plugins/system/cookiemanager/cookiemanager.php]:
```suggestion
		$db    = Factory::getDbo();
		$query = $db->getQuery(true)
			->select($db->quoteName(['id','title','alias','description']))
			->from($db->quoteName('#__categories'))
			->where([
				$db->quoteName('extension') . ' = ' . $db->quote('com_cookiemanager'),
				$db->quoteName('published') . ' =  1',
				]
			)
			->order($db->quoteName('lft'));

		$db->setQuery($query);
		$category = $db->loadObjectList();

		$bannerBody = '<p>' . Text::_('COM_COOKIEMANAGER_COOKIE_BANNER_DESCRIPTION') . '</p><p><a '
			. ' href="' . $menuitem->link . '">' . Text::_('COM_COOKIEMANAGER_VIEW_COOKIE_POLICY') . '</a></p>'
			. '<h5>' . Text::_('COM_COOKIEMANAGER_MANAGE_CONSENT_PREFERENCES') . '</h5><ul>';

```

File: plugins/system/cookiemanager/cookiemanager.php [hunk 7]
```diff
@@ -0,0 +1,222 @@
...
+					'RIGHT',
+					$db->quoteName('#__cookiemanager_cookies', 'a') . ' ON ' . $db->quoteName('c.id') . ' = ' . $db->quoteName('a.catid') . 'AND' . $db->quoteName('a.published') . ' =  1'
+				)
+				->order($db->quoteName('lft'));
+
+			$db->setQuery($query);
+			$cookies = $db->loadObjectList();
+
+			$prefBody = '<p>' . Text::_('COM_COOKIEMANAGER_PREFERENCES_DESCRIPTION') . '</p>';
+			$prefBody .= '<p><a  href="' . $menuitem->link . '">' . Text::_('COM_COOKIEMANAGER_VIEW_COOKIE_POLICY') . '</a></p>';
```
Comment (richard67) [inline plugins/system/cookiemanager/cookiemanager.php]:
```suggestion
		$db = $this->db;
		$query = $db->getQuery(true)
			->select($db->quoteName(['c.id','c.alias','a.cookie_name','a.cookie_desc','a.exp_period','a.exp_value']))
			->from($db->quoteName('#__categories', 'c'))
			->join(
				'RIGHT',
				$db->quoteName('#__cookiemanager_cookies', 'a') . ' ON ' . $db->quoteName('c.id') . ' = ' . $db->quoteName('a.catid') . 'AND' . $db->quoteName('a.published') . ' =  1'
			)
			->order($db->quoteName('lft'));

		$db->setQuery($query);
		$cookies = $db->loadObjectList();

		$prefBody = '<p>' . Text::_('COM_COOKIEMANAGER_PREFERENCES_DESCRIPTION') . '</p>';
		$prefBody .= '<p><a  href="' . $menuitem->link . '">' . Text::_('COM_COOKIEMANAGER_VIEW_COOKIE_POLICY') . '</a></p>';
```

File: plugins/system/cookiemanager/cookiemanager.php [hunk 8]
```diff
@@ -0,0 +1,222 @@
...
+				[
+						'title' => Text::_('COM_COOKIEMANAGER_PREFERENCES_TITLE'),
+						'footer' => '<button type="button" class="btn btn-info" data-bs-toggle="modal" data-bs-target="#preferences">'
+						. Text::_('COM_COOKIEMANAGER_CONFIRM_CHOICE_BUTTON_TEXT') . '</button>'
+						. '<button type="button" class="btn btn-info" data-bs-toggle="modal" data-bs-target="#preferences">'
+						. Text::_('COM_COOKIEMANAGER_ACCEPT_BUTTON_TEXT') . '</button>'
+
+					],
+				$prefBody
+			);
```
Comment (richard67) [inline plugins/system/cookiemanager/cookiemanager.php]:
```suggestion
		$this->preferences = HTMLHelper::_(
			'bootstrap.renderModal',
			'preferences',
			[
				'title' => Text::_('COM_COOKIEMANAGER_PREFERENCES_TITLE'),
				'footer' => '<button type="button" class="btn btn-info" data-bs-toggle="modal" data-bs-target="#preferences">'
					. Text::_('COM_COOKIEMANAGER_CONFIRM_CHOICE_BUTTON_TEXT') . '</button>'
					. '<button type="button" class="btn btn-info" data-bs-toggle="modal" data-bs-target="#preferences">'
					. Text::_('COM_COOKIEMANAGER_ACCEPT_BUTTON_TEXT') . '</button>'
			],
			$prefBody
		);

```

File: plugins/system/cookiemanager/cookiemanager.php [hunk 9]
```diff
@@ -0,0 +1,342 @@
...
+ * System plugin to manage cookies.
+ *
+ * @since  __DEPLOY_VERSION__
+ */
+class PlgSystemCookiemanager extends CMSPlugin
+{
+	/**
+	 * Application object.
+	 *
+	 * @var    JApplicationCms
```
Comment (bembelimen) [inline plugins/system/cookiemanager/cookiemanager.php]: No "J" classes

File: plugins/system/cookiemanager/cookiemanager.php [hunk 10]
```diff
@@ -0,0 +1,342 @@
...
+	 *
+	 * @return  void
+	 *
+	 * @since   __DEPLOY_VERSION__
+
+	 */
+	public function onBeforeCompileHead()
+	{
+		ob_start();
+		ob_implicit_flush(false);
```
Comment (bembelimen) [inline plugins/system/cookiemanager/cookiemanager.php]: Why is this needed?

File: plugins/system/cookiemanager/cookiemanager.php [hunk 11]
```diff
@@ -0,0 +1,342 @@
...
+
+		if (!$this->app->isClient('site'))
+		{
+			return;
+		}
+
+		$this->app->getDocument()->getWebAssetManager()
+			->registerAndUseScript('plg_system_cookiemanager.script', 'plg_system_cookiemanager/cookiemanager.min.js', [], ['defer' => true], ['core'])
+			->registerAndUseStyle('plg_system_cookiemanager.style', 'plg_system_cookiemanager/cookiemanager.min.css');
```
Comment (bembelimen) [inline plugins/system/cookiemanager/cookiemanager.php]: probably you can move this in a "joomla.asset.json" file?

File: plugins/system/cookiemanager/cookiemanager.php [hunk 12]
```diff
@@ -0,0 +1,342 @@
...
+
+		$lang = Factory::getLanguage();
+		$lang->load('com_cookiemanager', JPATH_ADMINISTRATOR);
```
Comment (bembelimen) [inline plugins/system/cookiemanager/cookiemanager.php]: This could be moved into the ```loadLanguage``` method.

File: plugins/system/cookiemanager/cookiemanager.php [hunk 13]
```diff
@@ -0,0 +1,342 @@
...
+		Text::script('COM_COOKIEMANAGER_PREFERENCES_LESS_BUTTON_TEXT');
+		Text::script('COM_COOKIEMANAGER_PREFERENCES_MORE_BUTTON_TEXT');
+
+		$params = ComponentHelper::getParams('com_cookiemanager');
+		$sitemenu = $this->app->getMenu();
+		$menuitem = $sitemenu->getItem($params->get('policylink'));
+
+		$config = [];
+		$config['expiration'] = $params->get('consent_expiration');
+		$config['position'] = $params->get('modal_position');
```
Comment (bembelimen) [inline plugins/system/cookiemanager/cookiemanager.php]: Params values should have default values, if they're not set.

File: plugins/system/cookiemanager/cookiemanager.php [hunk 14]
```diff
@@ -0,0 +1,342 @@
...
+		$this->app->getDocument()->addScriptOptions('config', $config);
```
Comment (bembelimen) [inline plugins/system/cookiemanager/cookiemanager.php]: Probably use a better name, which has a prefix from the plugin ("config" is very generic and could be conflicted with something else?)

File: plugins/system/cookiemanager/cookiemanager.php [hunk 15]
```diff
@@ -0,0 +1,342 @@
...
+			. ' href="' . $menuitem->link . '">' . Text::_('COM_COOKIEMANAGER_VIEW_COOKIE_POLICY') . '</a></p>'
+			. '<h5>' . Text::_('COM_COOKIEMANAGER_MANAGE_CONSENT_PREFERENCES') . '</h5><ul>';
+
+		foreach ($this->category as $key => $value)
+		{
+			$bannerBody .= '<li class="cookie-cat form-check form-check-inline"><label>' . $value->title . '<span class="ms-4 form-check-inline form-switch"><input class="form-check-input" data-cookiecategory="'
+			. $value->alias . '" type=checkbox></span></label></li>';
+		}
+
+		$bannerBody .= '</ul>';
```
Comment (bembelimen) [inline plugins/system/cookiemanager/cookiemanager.php]: Move HTML code to tmpl files, so they can be overriden.

File: plugins/system/cookiemanager/cookiemanager.php [hunk 16]
```diff
@@ -0,0 +1,342 @@
...
+		$this->cookieBanner = HTMLHelper::_(
+			'bootstrap.renderModal',
+			'cookieBanner',
+			[
+					'title' => Text::_('COM_COOKIEMANAGER_COOKIE_BANNER_TITLE'),
+					'footer' => '<button type="button" id="bannerConfirmChoice" class="btn btn-info" data-bs-dismiss="modal">'
+					. Text::_('COM_COOKIEMANAGER_CONFIRM_CHOICE_BUTTON_TEXT') . '</button>'
+					. '<button type="button" class="btn btn-info" data-bs-toggle="modal" data-bs-dismiss="modal" data-bs-target="#preferences">'
+					. Text::_('COM_COOKIEMANAGER_MORE_DETAILS') . '</button>'
+					. '<button type="button" "btnAcceptAllCookies" class="btn btn-info" data-bs-dismiss="modal">'
```
Comment (bembelimen) [inline plugins/system/cookiemanager/cookiemanager.php]: ```type="button" "btnAcceptAllCookies" class="btn btn-info"```

Here is something broken...

File: plugins/system/cookiemanager/cookiemanager.php [hunk 17]
```diff
@@ -0,0 +1,342 @@
...
+			->select($db->quoteName(['c.id','c.alias','a.cookie_name','a.cookie_desc','a.exp_period','a.exp_value']))
+			->from($db->quoteName('#__categories', 'c'))
+			->join(
+				'RIGHT',
+				$db->quoteName('#__cookiemanager_cookies', 'a') . ' ON ' . $db->quoteName('c.id') . ' = ' . $db->quoteName('a.catid') . 'AND' . $db->quoteName('a.published') . ' =  1'
+			)
+			->order($db->quoteName('lft'));
+
+		$db->setQuery($query);
+		$cookies = $db->loadObjectList();
```
Comment (bembelimen) [inline plugins/system/cookiemanager/cookiemanager.php]: You can concatinate the two calls.

File: plugins/system/cookiemanager/cookiemanager.php [hunk 18]
```diff
@@ -0,0 +1,342 @@
...
+						. '<td>' . $value->cookie_desc . '</td>'
+						. '<td>' . $value->exp_value . ' ' . $value->exp_period . '</td>'
+						. '</tr>';
+					}
+				}
+			}
+
+			$table .= '</tbody></table>';
+			$prefBody .= $table . '</div>';
+		}
```
Comment (bembelimen) [inline plugins/system/cookiemanager/cookiemanager.php]: Move HTML to tmpl file

File: plugins/system/cookiemanager/cookiemanager.php [hunk 19]
```diff
@@ -0,0 +1,342 @@
...
+				{
+					if ($catValue->id == $value->catid)
+					{
+						array_push($this->script[$catValue->alias], $value);
+					}
+				}
+			}
+		}
+
+				$this->app->getDocument()->addScriptOptions('code', $this->script);
```
Comment (bembelimen) [inline plugins/system/cookiemanager/cookiemanager.php]: Naming again should be plugin related.

File: plugins/system/cookiemanager/cookiemanager.php [hunk 20]
```diff
@@ -0,0 +1,342 @@
...
+			return;
+		}
+
+		echo $this->cookieBanner;
+		echo $this->preferences;
+		echo '<button class="preview btn btn-info" data-bs-toggle="modal" data-bs-target="#cookieBanner">' . Text::_('COM_COOKIEMANAGER_PREVIEW_BUTTON_TEXT') . '</button>';
+
+		foreach ($this->category as $catKey => $catValue)
+		{
+			if (isset($_COOKIE['cookie_category_' . $catValue->alias]) && $_COOKIE['cookie_category_' . $catValue->alias] === 'true')
```
Comment (bembelimen) [inline plugins/system/cookiemanager/cookiemanager.php:281]: ```$this->app->input->cookie->set(...)```

File: plugins/system/cookiemanager/cookiemanager.php [hunk 21]
```diff
@@ -0,0 +1,342 @@
...
+			{
+				$this->script[$catValue->alias] = [];
+
+				foreach ($this->scripts as $key => $value)
```
Comment (bembelimen) [inline plugins/system/cookiemanager/cookiemanager.php]: This foreach looks a bit "smelly", I think this has a potential to break...we have to check.

File: plugins/system/cookiemanager/cookiemanager.php (hunk 10 above)
Comment (rjharishabh) [reply plugins/system/cookiemanager/cookiemanager.php]: Because we need to render cookie setting scripts 
So first capture the output buffer, concatenate with scripts and then render it (get idea from debug plugin)

File: plugins/system/cookiemanager/cookiemanager.php (hunk 18 above)
Comment (rjharishabh) [reply plugins/system/cookiemanager/cookiemanager.php]: Done

File: plugins/system/cookiemanager/cookiemanager.php (hunk 15 above)
Comment (rjharishabh) [reply plugins/system/cookiemanager/cookiemanager.php]: Done
